sensor/kubernetes/certrefresh/securedcluster_tls_issuer.go

```diff
@@ -117,22 +117,23 @@ func (i *securedClusterTLSIssuerImpl) ResponsesC() <-chan *message.ExpiringMessa
 // This method must not block as it would prevent centralReceiverImpl from sending messages
 // to other SensorComponents.
 func (i *securedClusterTLSIssuerImpl) ProcessMessage(msg *central.MsgToSensor) error {
-	switch m := msg.GetMsg().(type) {
-	case *central.MsgToSensor_IssueSecuredClusterCertsResponse:
-		response := m.IssueSecuredClusterCertsResponse
+	if m, ok := msg.GetMsg().(*central.MsgToSensor_IssueSecuredClusterCertsResponse); ok {
 		go func() {
-			ctx, cancel := context.WithTimeout(context.Background(), processMessageTimeout)
-			defer cancel()
-			select {
-			case <-ctx.Done():
-				// certRefresher will retry.
-				log.Errorf("timeout forwarding response %s from central: %s", response, ctx.Err())
-			case i.msgFromCentralC <- response:
-			}
+			i.certRequester.DispatchResponse(certificates.NewResponseFromSecuredClusterCerts(m.IssueSecuredClusterCertsResponse))
 		}()
+	}
+
+	// messages not supported by this component are ignored
+	return nil
+}
+
+func (i *securedClusterTLSIssuerImpl) msgToCentralHandler(ctx context.Context, msg *message.ExpiringMessage) error {
```

Comment:
Okay, this is the sneaky and tricky part. Here are the issues that I could identify.

1. This function is used to allow other objects write into a channel that is created somewhere else. This makes it impossible to close that channel when no more messages will be created on it. This is not a bug but an antipattern that will make creating bugs easier in the future.
2. If the `i.stopSig.Done` triggers, then all writers that will call this function will block. There is no way of informing the writers that they should stop producing messages.


I will refer to this issue in other comments as "the blocker" 🛑.

Reply:
1. I was thinking this might be acceptable, because the TLSIssuer owns the Requester (which writes in the TLSIssuer's channel using this callback), and the Requester is the only object that needs to write into the TLSIssuer's channel. I could add back a `Stop` method to the Requester could stop ongoing requests, and call it from TLSIssuer.

2. Is that really so? I think that if `i.stopSig.Done()` is triggered (i.e. the channel is closed), then `msgToCentralHandler` will just exit immediately and that's fine.